balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,627 @@
+	ch := make(chan struct{})
+	b.serializer.ScheduleOr(func(ctx context.Context) {
+		b.exitIdle()
+		close(ch)
+	}, func() {
+		close(ch)
+	})
+	<-ch
+}
+
+func (b *pickfirstBalancer) exitIdle() {
+	b.requestConnection()
+}
+
+// deDupAddresses ensures that each address belongs to only one endpoint.
+func deDupAddresses(endpoints []resolver.Endpoint) []resolver.Endpoint {
+	seenAddrs := resolver.NewAddressMap()
+	newEndpoints := []resolver.Endpoint{}
+
+	for _, ep := range endpoints {
+		addrs := []resolver.Address{}
+		for _, addr := range ep.Addresses {
+			if _, ok := seenAddrs.Get(addr); ok {
+				continue
+			}
+			addrs = append(addrs, addr)
+		}
+		if len(addrs) == 0 {
+			continue
+		}
+		newEndpoints = append(newEndpoints, resolver.Endpoint{
+			Addresses:  addrs,
+			Attributes: ep.Attributes,
+		})
+	}
+	return newEndpoints
+}
+
+// shutdownRemaining shuts down remaining subConns. Called when a subConn
+// becomes ready, which means that all other subConn must be shutdown.
+func (b *pickfirstBalancer) shutdownRemaining(selected *scData) {
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.subConn != selected.subConn {
+			sd.subConn.Shutdown()
+		}
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+	b.subConns.Set(*selected.addr, selected)
+}
+
+// requestConnection requests a connection to the next applicable address'
+// subcon, creating one if necessary. Schedules a connection to next address in list as well.
+// If the current channel has already attempted a connection, we attempt a connection
+// to the next address/subconn in our list.  We assume that NewSubConn will never
+// return an error.
+func (b *pickfirstBalancer) requestConnection() {
+	if !b.addressIndex.isValid() || b.state == connectivity.Shutdown {
+		return
+	}
+	curAddr, err := b.addressIndex.currentAddress()
+	if err != nil {
+		// This should not never happen because we already check for validity and
+		// return early above.
+		return
+	}
+	sd, ok := b.subConns.Get(*curAddr)
+	if !ok {
+		sd, err = newScData(b, *curAddr)
+		if err != nil {
+			// This should never happen.
+			b.logger.Warningf("Failed to create a subConn for address %v: %v", curAddr.String(), err)
+			b.state = connectivity.TransientFailure
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("error creating connection: %v", err)},
+			})
+			b.addressIndex.reset()
+			return
+		}
+		b.subConns.Set(*curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressIndex.increment() {
+			b.endFirstPass()
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Warningf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, state balancer.SubConnState) {
```

Comment:
It would be good to document methods that are expected to be run *only* in seralizer callbacks in their docstring. Something like this:
```
// methodFoo does blah and foo.
//
// Only executed in the context of a serializer callback.
```